### PR TITLE
fix todo [tls cache]

### PR DESCRIPTION
**Pull request by ls-2018**
fix 
```// TODO: cache```

### CODE REVIEW

**Review by ptabor — APPROVED** [review]
Thank you

**Review by serathius — CHANGES_REQUESTED** [review]
I know there is a todo, but I don't think we should assume that the author of it has really though through that cache here is needed. Correctly implemented cache allows us to trade Disk IO & cpu for memory. However it's not always obvious that cache is beneficial, depending on hit vs miss ratio it might lead to lowering performance or resource leaking.

There are two things we should definetly do before merging:
* Add tests
* Guarantee maximum size of cache (implement LRU cache)

**Comment by ptabor** [comment]
> I know there is a todo, but I don't think we should assume that the author of it has really though through that cache here is needed. Correctly implemented cache allows us to trade Disk IO & cpu for memory. However it's not always obvious that cache is beneficial, depending on hit vs miss ratio it might lead to lowering performance or resource leaking.
> 
> There are two things we should definetly do before merging:
> 
> * Add tests
> * Guarantee maximum size of cache (implement LRU cache)

I agree that LRU based implementation (ideally with O(1min)) expiration would be better. 

Few observations: 
 1. I think this optimisation might be pretty important in case we have a lot of new connections.
     The current model assumes re-read of the certificate from file when each connection is being created: 
      https://github.com/etcd-io/etcd/blob/0c9a4e0f93ad9a7d83e8f5ec215baa856616c76b/client/pkg/transport/listener_tls.go#L155
    (and not what intuition says: "when listener is being created"). 
   If we have storm of connections, e.g. after server restart... it might participate significantly both: in terms of time and memory when we need to reread the certs. 

 2. This code is in `client/pkg`, but it's about opening 'listeners' that is used only on server side, e.g. https://github.com/etcd-io/etcd/blob/0c9a4e0f93ad9a7d83e8f5ec215baa856616c76b/server/embed/etcd.go#L530-L534.  So in practice the configuration of the file-names does not change frequently/ever so in practical application I expect 'fixed' O(1-4) certs being loaded. 
 In theory some application of `embed` server might keep creating the server's within a single process with different configs... potentially over-feeling the cache. 

3. There is aspect of 'impact' of substituting a file in place (while server is running) with 'refreshed' certificate. In theory it used to be 'working' before the PR, and with this change, the server would stick to the original content. Thus invalidating the cache with O(1-10 min) min resolution would be ideal. 

Summary: 
  - Introducing caching is IMHO justified. Proving an impact on perf during connection storm would be an appreciated exercise. 
  - Using some existing lro implementation (hashicorp ?) would make the code safer and is recommended.

**Comment by ls-2018** [comment]
I agree with the LRU based implementation and I will implement it next

**Review by ptabor — CHANGES_REQUESTED** [review]
Thank you for the fix. 

I wonder whether we can avoid maintaining own implementation of LRU. 
E.g. usage of https://github.com/hashicorp/golang-lru with a wrapper that 
substitutes expired entries might be suffient and imply less maintainance cost. 

I've seen also `https://github.com/hnlq715/golang-lru`, but it's a fork without it's own brand, so would be misleading to depend on it.

File: client/pkg/lruutil/lru_cache.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
+// Copyright 2022 Google LLC
```
Comment (ptabor) [inline client/pkg/lruutil/lru_cache.go:1]: I assume it's some copy-paste problem. 

Usually we use: `// Copyright 2022 The etcd Authors`

**Comment by ls-2018** [comment]
@ptabor  I've seen https://github.com/hnlq715/golang-lru. but it's not support time-based elimination strategies

**Comment by ptabor** [comment]
> @ptabor I've seen https://github.com/hnlq715/golang-lru. but it's not support time-based elimination strategies.

If we take any existing LRO, we can simulate expiration on fetch.

```
	if v, ok := crlBytesLru.Get(crlPath); ok {
		if v.expiration.Before(time.now()) {
		  return crlBytes.value
		}
	} 
	crlBytes, err := ioutil.ReadFile(crlPath)
	if err != nil {
		return err
	}
	crlBytesLru.Set(crlPath, expirable{value:crlBytes, expiration:Time.now()+ttl})
	}

```
Even better wrapped in a function: 

     memoize(lru, crlPath, ttl, func() { return ioutil.ReadFile(crlPath) } )

> action fails just by changing a copyright
likely flake. Restarted.

**Pull request closed without merging**